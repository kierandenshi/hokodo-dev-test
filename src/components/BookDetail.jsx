import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { selectBookById } from '@redux/modules/books';
import { Component as BookList } from '@components/BookList';
import BookCover from '@components/BookCover';

const StyledItem = styled.div`
  margin: 0 0 1rem;

  .title {
    font-size: 2rem;
    font-weight: bold;
  }

  .author {
    font-size: 1.5rem;
  }
`;

const StyledLink = styled(Link)`
  color: green;
`;

const BookDetail = ({ book: { id, name, author, isbn, cover, byAuthor } }) =>
  id ? (
    <Fragment>
      <StyledItem>
        <BookCover imageUrl={cover} size="full" />
        <div className="title">{name}</div>
        <div className="author">{author.name}</div>
        <div className="isbn">{`ISBN: ${isbn}`}</div>
      </StyledItem>
      {byAuthor.length ? (
        <Fragment>
          <div>More by this author:</div>
          <BookList books={byAuthor} />
        </Fragment>
      ) : null}
      <StyledLink to="/">Back to full book list</StyledLink>
    </Fragment>
  ) : null;

BookDetail.propTypes = {
  book: PropTypes.shape({}).isRequired,
};

const Container = connect((state, props) => ({
  book: selectBookById(state, { id: props.match.params.id }),
}))(BookDetail);

export { Container as default, BookDetail as Component };
