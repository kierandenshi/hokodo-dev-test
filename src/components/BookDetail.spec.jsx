import React from 'react';
import { shallow } from 'enzyme';
import { Component } from './BookDetail';

const defaultProps = {
  book: {
    id: 'book1',
    name: 'Really Good Book',
    author: {
      name: 'Mr. John Doe',
      id: 'author1',
    },
    isbn: '1234',
    byAuthor: [],
    cover: 'http://coverimage.url',
  },
};

const getWrapper = (testProps = {}) => {
  const props = { ...defaultProps, ...testProps };
  return shallow(<Component {...props} />);
};

describe('Component', () => {
  it('should match snapshot', () => {
    const wrapper = getWrapper();
    expect(wrapper).toMatchSnapshot();
  });

  it('should match snapshot', () => {
    const testProps = {
      book: {
        id: 'book2',
        author: {
          id: 'author2',
          name: 'Prof. John Doe',
        },
        name: 'Schuster Champlinborough',
        isbn: '1441614311',
        published_at: '1991-07-29',
        cover: 'https://lorempixel.com/640/480/?84662',
        byAuthor: [
          {
            id: 'book3',
            author: {
              id: 'author2',
              name: 'Prof. John Doe',
            },
            name: 'Okuneva MarianoVille',
            isbn: '0779284704',
            published_at: '2017-06-30',
            cover: 'https://lorempixel.com/640/480/?28970',
          },
        ],
      },
    };
    const wrapper = getWrapper(testProps);
    expect(wrapper).toMatchSnapshot();
  });
});
