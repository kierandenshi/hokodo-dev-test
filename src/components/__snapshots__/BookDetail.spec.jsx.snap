// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component should match snapshot 1`] = `
<Fragment>
  <styled.div>
    <BookCover
      imageUrl="http://coverimage.url"
      size="full"
    />
    <div
      className="title"
    >
      Really Good Book
    </div>
    <div
      className="author"
    >
      Mr. John Doe
    </div>
    <div
      className="isbn"
    >
      ISBN: 1234
    </div>
  </styled.div>
  <Styled(Link)
    to="/"
  >
    Back to full book list
  </Styled(Link)>
</Fragment>
`;

exports[`Component should match snapshot 2`] = `
<Fragment>
  <styled.div>
    <BookCover
      imageUrl="https://lorempixel.com/640/480/?84662"
      size="full"
    />
    <div
      className="title"
    >
      Schuster Champlinborough
    </div>
    <div
      className="author"
    >
      Prof. John Doe
    </div>
    <div
      className="isbn"
    >
      ISBN: 1441614311
    </div>
  </styled.div>
  <div>
    More by this author:
  </div>
  <BookList
    books={
      Array [
        Object {
          "author": Object {
            "id": "author2",
            "name": "Prof. John Doe",
          },
          "cover": "https://lorempixel.com/640/480/?28970",
          "id": "book3",
          "isbn": "0779284704",
          "name": "Okuneva MarianoVille",
          "published_at": "2017-06-30",
        },
      ]
    }
  />
  <Styled(Link)
    to="/"
  >
    Back to full book list
  </Styled(Link)>
</Fragment>
`;
